core/state/snapshot/generate.go

```diff
@@ -404,16 +406,31 @@ func (dl *diskLayer) generateRange(root common.Hash, prefix []byte, kind string,
 		}
 		meter.Mark(1)
 	}
+
+	// We use the snap data to build up a cache which can be used by the
+	// main account trie as a primary lookup when resolving hashes
+	var snapTrieDb *trie.Database
+	if len(result.keys) > 0 {
+		snapNodeCache := memorydb.New()
+		snapTrieDb = trie.NewDatabase(snapNodeCache)
+		snapTrie, _ := trie.New(common.Hash{}, snapTrieDb)
+		for i, key := range result.keys {
+			snapTrie.Update(key, result.vals[i])
+		}
+		snapTrie.Commit(nil)
+	}
 	tr := result.tr
 	if tr == nil {
 		tr, err = trie.New(root, dl.triedb)
 		if err != nil {
 			return false, nil, err
 		}
 	}
+
 	var (
 		trieMore       bool
-		iter           = trie.NewIterator(tr.NodeIterator(origin))
+		nodeIt         = tr.NodeIterator(origin)
```

Comment:
I think we can pass the auxiliary database here. `iter = trie.NewIterator(tr.NodeIteratorWithDatabase(origin))`